dll/win32/shlwapi/autocomp.c

```diff
@@ -0,0 +1,707 @@
+    BSTR bstr, *pstrs;
+
+    bstr = SysAllocString(str);
+    if (!bstr)
+        return FALSE;
+
+    if (this->m_cstrs + 1 >= this->m_capacity)
+    {
+        new_capacity = this->m_capacity + NUM_GROW;
+        pstrs = (BSTR *)CoTaskMemAlloc(new_capacity * sizeof(BSTR));
+        if (!pstrs)
+        {
+            SysFreeString(bstr);
+            return FALSE;
+        }
+
+        CopyMemory(pstrs, this->m_pstrs, this->m_cstrs * sizeof(BSTR));
+        CoTaskMemFree(this->m_pstrs);
+        this->m_pstrs = pstrs;
+        this->m_capacity = new_capacity;
+    }
+
+    this->m_pstrs[this->m_cstrs++] = bstr;
+    return TRUE;
+}
+
+static HRESULT STDMETHODCALLTYPE
+AC_EnumString_Next(
+    IEnumString* This,
+    ULONG celt,
+    LPOLESTR *rgelt,
+    ULONG *pceltFetched)
+{
+    SIZE_T ielt, cch, cb;
+    BSTR *pstrs;
+    AC_EnumString *this = (AC_EnumString *)This;
+
+    if (!rgelt || !pceltFetched)
+        return E_POINTER;
+
+    *pceltFetched = 0;
+    *rgelt = NULL;
+
+    if (this->m_istr >= this->m_cstrs)
+        return S_FALSE;
+
+    pstrs = this->m_pstrs;
+    for (ielt = 0;
+         ielt < celt && this->m_istr < this->m_cstrs;
+         ++ielt, ++this->m_istr)
+    {
+        cch = (SysStringLen(pstrs[this->m_istr]) + 1);
+        cb = cch * sizeof(WCHAR);
+
+        rgelt[ielt] = (LPWSTR)CoTaskMemAlloc(cb);
+        if (!rgelt[ielt])
+        {
+            while (ielt-- > 0)
+            {
+                CoTaskMemFree(rgelt[ielt]);
+            }
+            return S_FALSE;
+        }
+        CopyMemory(rgelt[ielt], pstrs[this->m_istr], cb);
+    }
+
+    *pceltFetched = ielt;
+
+    if (ielt == celt)
+        return S_OK;
+
+    return S_FALSE;
+}
+
+static HRESULT STDMETHODCALLTYPE
+AC_EnumString_Skip(IEnumString* This, ULONG celt)
+{
+    AC_EnumString *this = (AC_EnumString *)This;
+    if (this->m_istr + celt >= this->m_cstrs || 0 == this->m_cstrs)
+        return S_FALSE;
+
+    this->m_istr += celt;
+    return S_OK;
+}
+
+static AC_EnumString *AC_EnumString_Construct(SIZE_T capacity);
+
+static HRESULT STDMETHODCALLTYPE
+AC_EnumString_Clone(IEnumString* This, IEnumString **ppenum)
+{
+    AC_EnumString *this, *cloned;
+    SIZE_T i, count;
+    BSTR *pstrs;
+
+    if (!ppenum)
+        return E_POINTER;
+
+    this = (AC_EnumString *)This;
+    cloned = AC_EnumString_Construct(this->m_cstrs);
+    if (!cloned)
+        return E_OUTOFMEMORY;
+
+    count = this->m_cstrs;
+    pstrs = this->m_pstrs;
+    for (i = 0; i < count; ++i)
+    {
+        if (!AC_EnumString_AddBStrNoGrow(cloned, pstrs[i]))
+        {
+            AC_EnumString_Destruct(cloned);
+            return E_FAIL;
+        }
+    }
+
+    cloned->m_capacity = cloned->m_cstrs = count;
+    cloned->m_hwndEdit = this->m_hwndEdit;
+    cloned->m_dwSHACF_ = this->m_dwSHACF_;
+
+    *ppenum = (IEnumString *)cloned;
+    return S_OK;
+}
+
+#define IS_IGNORED_DOTS(sz) ( \
+    sz[0] == L'.' && (sz[1] == 0 || (sz[1] == L'.' && sz[2] == 0)) \
+)
```

Comment:
You can save a new line by putting the first opening & final closing parentheses on the same line as the tests on sz[0] and sz[1] you do:
```c
#define ARE_DOTS_IGNORED(sz) \
    ( sz[0] == L'.' && (sz[1] == 0 || (sz[1] == L'.' && sz[2] == 0)) )
```